relengapi/blueprints/slaveloan/__init__.py

```diff
@@ -0,0 +1,233 @@
+
+
+@bp.route('/loans/request', methods=['POST'])
+@apimethod(None, body=rest.LoanRequest)
+def new_loan_request(body):
+    "User Loan Requesting, returns the id of the loan"
+    if not body.ldap_email:
+        raise BadRequest("Missing LDAP E-Mail")
+    if not body.requested_slavetype:
+        raise BadRequest("Missing slavetype")
+
+    slavetype = slave_to_slavetype(body.requested_slavetype)
+    if not slavetype:
+        raise BadRequest("Unsupported slavetype")
+
+    if not body.bugzilla_email:
+        # Set bugzilla e-mail to ldap e-mail by default
+        body.bugzilla_email = body.ldap_email
+
+    session = g.db.session('relengapi')
+    try:
+        h = Humans.as_unique(session, ldap=body.ldap_email,
+                             bugzilla=body.bugzilla_email)
+    except sa.exc.IntegrityError:
+        raise InternalServerError("Integrity Error from Database, please retry.")
+
+    if body.loan_bug_id:
+        l = Loans(status="PENDING", human=h, bug_id=body.loan_bug_id)
+    else:
+        l = Loans(status="PENDING", human=h)
+
+    history = History(for_loan=l,
+                      timestamp=tz.utcnow(),
+                      msg="Requesting loan for slavetype %s (original: '%s')" %
+                          (slavetype, body.requested_slavetype))
+    session.add(l)
+    session.add(history)
+    session.commit()
+    chain_of_stuff = task_groups.generate_loan(loanid=l.id, slavetype=slavetype)
+    chain_of_stuff.delay()
+    return None  # ?rest.WSME_New_Loan({'loan': l.to_wsme()})
+
+
+@bp.route('/machine/classes')
+@apimethod(
+    wsme.types.DictType(
+        key_type=unicode,
+        value_type=wsme.types.ArrayType(unicode)
+    )
+)
+def get_machine_classes():
+    """
+    A mapping of what you'll get with a given loan, and globs of the slave types associated.
+
+    Returns a mapping keyed on type of loan against slave-name globs that it corresponds to
+    e.g.:::
+
+        {
+            "b-2008-ix": [
+                "b-2008-ix-*",
+                "b-2008-sm-*",
+                "w64-ix-*"
+            ],
+        }
+
+    Where the above would tell you we are loaning a 'b-2008-ix' machine for slaves
+    which match any of the globs in the array."""
+    return slave_patterns()
+
+##################
+# User Interface #
+##################
+
+
+@bp.route('/')
+@flask_login.login_required
+def root():
+    return render_template('slaveloan_root.html')
+
+
+@bp.route('/details/<int:id>')
+@flask_login.login_required
+@p.slaveloan.admin.require()
+def loan_details(id):
+    g.loanid = id
+    return render_template('slaveloan_details.html')
+
+
+@bp.route('/admin/')
+@flask_login.login_required
+@p.slaveloan.admin.require()
+def admin():
+    return render_template('slaveloan_admin.html')
+
+##################
+# Temporary      #
+##################
+
+
+@bp.route('/tmp/')
+@p.slaveloan.admin.require()
+def init_loan():
+    "Temporary, Manual testing only, DO NOT USE"
```

Comment:
ugh, nope --- temporary like, shouldn't even be in this patch. (I was using it to test celery task progress before I had a better way ;-) )
So removed :white_check_mark: